File: app/src/main/java/io/nekohasekai/sagernet/fmt/AbstractBean.java
package io.nekohasekai.sagernet.fmt;

import static io.nekohasekai.sagernet.ktx.MapsKt.getGson;

import androidx.annotation.NonNull;

import com.esotericsoftware.kryo.io.ByteBufferInput;
import com.esotericsoftware.kryo.io.ByteBufferOutput;

import io.nekohasekai.sagernet.ktx.NetsKt;

import org.jetbrains.annotations.NotNull;

import java.util.Arrays;

public abstract class AbstractBean extends Serializable {

    public String serverAddress;
    public Integer serverPort;
    public String name;
    public String customOutboundJson;
    public String customConfigJson;

    public Boolean serverMux;
    public Boolean serverBrutal;
    public Integer serverMuxType;
    public Integer serverMuxNumber;
    public Boolean serverMuxPadding;
    public Integer serverMuxStrategy;

    public transient String finalAddress;
    public transient int finalPort;
    private transient boolean serializeWithoutName;

    public String displayName() {
        if (name != null && !name.isEmpty()) {
            return name;
        } else {
            return displayAddress();
        }
    }

    public String displayAddress() {
        return NetsKt.wrapIPV6Host(serverAddress) + ":" + serverPort;
    }

    public String network() {
        return "tcp,udp";
    }

    public boolean canICMPing() {
        return true;
    }

    public boolean canTCPing() {
        return true;
    }

    public boolean needUDPOverTCP() {
        return false;
    }

    public boolean canMapping() {
        return true;
    }

    @Override
    public void initializeDefaultValues() {
        if (serverAddress == null || serverAddress.isEmpty()) {
            serverAddress = "127.0.0.1";
        } else if (serverAddress.startsWith("[") && serverAddress.endsWith("]")) {
            serverAddress = NetsKt.unwrapIPV6Host(serverAddress);
        }
        if (serverPort == null) {
            serverPort = 1080;
        }
        if (name == null) name = "";

        finalAddress = serverAddress;
        finalPort = serverPort;

        if (customOutboundJson == null) customOutboundJson = "";
        if (customConfigJson == null) customConfigJson = "";

        if (serverMux == null) serverMux = false;
        if (serverBrutal == null) serverBrutal = false;
        if (serverMuxType == null) serverMuxType = 0;
        if (serverMuxNumber == null) serverMuxNumber = 8;
        if (serverMuxPadding == null) serverMuxPadding = false;
        if (serverMuxStrategy == null) serverMuxStrategy = 0;
    }

    @Override
    public void serializeToBuffer(@NonNull ByteBufferOutput output) {
        serialize(output);

        output.writeInt(4);
        if (!serializeWithoutName) {
            output.writeString(name);
        }
        output.writeString(customOutboundJson);
        output.writeString(customConfigJson);

        output.writeBoolean(serverBrutal);
        output.writeBoolean(serverMux);
        output.writeInt(serverMuxType);
        output.writeInt(serverMuxNumber);
        output.writeBoolean(serverMuxPadding);
        output.writeInt(serverMuxStrategy);
    }

    @Override
    public void deserializeFromBuffer(@NonNull ByteBufferInput input) {
        deserialize(input);

        int extraVersion = input.readInt();

        name = input.readString();
        customOutboundJson = input.readString();
        customConfigJson = input.readString();

        if (extraVersion >= 2) serverBrutal = input.readBoolean();
        if (extraVersion < 3) return;
        serverMux = input.readBoolean();
        serverMuxType = input.readInt();
        serverMuxNumber = input.readInt();
        serverMuxPadding = input.readBoolean();
        if (extraVersion < 4) return;
        serverMuxStrategy = input.readInt();
    }

    public void serialize(ByteBufferOutput output) {
        output.writeString(serverAddress);
        output.writeInt(serverPort);
    }

    public void deserialize(ByteBufferInput input) {
        serverAddress = input.readString();
        serverPort = input.readInt();
    }

    @NotNull
    @Override
    public abstract AbstractBean clone();

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        try {
            serializeWithoutName = true;
            ((AbstractBean) o).serializeWithoutName = true;
            return Arrays.equals(KryoConverters.serialize(this), KryoConverters.serialize((AbstractBean) o));
        } finally {
            serializeWithoutName = false;
            ((AbstractBean) o).serializeWithoutName = false;
        }
    }

    @Override
    public int hashCode() {
        try {
            serializeWithoutName = true;
            return Arrays.hashCode(KryoConverters.serialize(this));
        } finally {
            serializeWithoutName = false;
        }
    }

    @NotNull
    @Override
    public String toString() {
        return getClass().getSimpleName() + " " + getGson().toJson(this);
    }

    public void applyFeatureSettings(AbstractBean other) {
    }

    @NotNull
    public String outboundType() throws Throwable {
        return "abstract";
    }

}

File: app/src/main/kotlin/io/nekohasekai/sagernet/fmt/AbstractBean.kt
package io.nekohasekai.sagernet.fmt

import android.text.TextUtils
import io.nekohasekai.sagernet.ktx.unwrapIPV6Host
import io.nekohasekai.sagernet.ktx.wrapIPV6Host
import com.esotericsoftware.kryo.io.ByteBufferInput
import com.esotericsoftware.kryo.io.ByteBufferOutput
import io.nekohasekai.sagernet.ktx.gson

abstract class AbstractBean : Serializable() {

    open val defaultPort = 1080

    var serverAddress: String = "127.0.0.1"
    var serverPort: Int = defaultPort
    var name: String = ""
    var customOutboundJson: String = ""
    var customConfigJson: String = ""
    var serverMux: Boolean = false
    var serverBrutal: Boolean = false
    var serverMuxType: Int = 0
    var serverMuxNumber: Int = 8
    var serverMuxPadding: Boolean = false
    var serverMuxStrategy: Int = 0
    var finalAddress: String = ""
    var finalPort: Int = 0

    open fun displayName(): String {
        if (TextUtils.isEmpty(name)) {
            return displayAddress()
        }
        return name
    }

    open fun displayAddress(): String {
        return "${serverAddress.wrapIPV6Host()}:$serverPort"
    }

    open fun network(): String {
        return "tcp,udp"
    }

    open val canICMPing = true
    open val canTCPing = true
    open val needUDPOverTCP = false
    open val canMapping = true

    override fun initializeDefaultValues() {
        if (serverAddress.isEmpty()) {
            serverAddress = "127.0.0.1"
        } else if (serverAddress.startsWith("[") && serverAddress.endsWith("]")) {
            serverAddress = serverAddress.unwrapIPV6Host()
        }
        if (serverPort !in 1..65535) {
            serverPort = defaultPort
        }
        finalAddress = serverAddress
        finalPort = serverPort
    }

    override fun serializeToBuffer(output: ByteBufferOutput) {
        serialize(output)

        output.writeInt(4)
        output.writeString(name)
        output.writeString(customOutboundJson)
        output.writeString(customConfigJson)

        output.writeBoolean(serverBrutal)
        output.writeBoolean(serverMux)
        output.writeInt(serverMuxType)
        output.writeInt(serverMuxNumber)
        output.writeBoolean(serverMuxPadding)
        output.writeInt(serverMuxStrategy)
    }

    override fun deserializeFromBuffer(input: ByteBufferInput) {
        deserialize(input)

        val extraVersion = input.readInt()

        name = input.readString()
        customOutboundJson = input.readString()
        customConfigJson = input.readString()

        if (extraVersion >= 2) {
            serverBrutal = input.readBoolean()
        }
        if (extraVersion >= 3) {
            serverMux = input.readBoolean()
            serverMuxType = input.readInt()
            serverMuxNumber = input.readInt()
            serverMuxPadding = input.readBoolean()
        }
        if (extraVersion >= 4) {
            serverMuxStrategy = input.readInt()
        }
    }

    open fun serialize(output: ByteBufferOutput) {
        output.writeString(serverAddress)
        output.writeInt(serverPort)
    }

    open fun deserialize(input: ByteBufferInput) {
        serverAddress = input.readString() ?: ""
        serverPort = input.readInt()
    }

    abstract fun clone(): AbstractBean

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        return KryoConverters.serialize(this)
            .contentEquals(KryoConverters.serialize(other as AbstractBean))
    }

    override fun hashCode(): Int {
        return KryoConverters.serialize(this).contentHashCode()
    }

    override fun toString(): String {
        return "${javaClass.simpleName} ${gson.toJson(this)}"
    }

    open fun applyFeatureSettings(other: AbstractBean) {}

}
